<launch>
  <arg name="image_topic" description="Raw image topic (currently we do not support rectified images)"/>
  <arg name="camera_info_topic"/>
  <arg name="pointcloud_topic"/>
  <arg name="base_frame" default="base_link"/>

  <arg name="use_receive_time" default="false" description="If sensor synchronization does not work, we can force the use of system time"/>
  <arg name="use_rectified_image" default="true" description="We need to know if the image has been rectified so we can determine the use of K or P"/>
  <arg name="calibration_pairs" default="9" description="Number of lidar-image pairs for the calibration to converge"/>
  <arg name="calibration_pairs_min_distance" default="1.5" description="Minimum allowed between a new detection and current pairs"/>

  <arg name="lidar_model">
    <choice value="aeva_aeries2"/>
    <choice value="at128"/>
    <choice value="innoviz2"/>
    <choice value="ouster_os1_128"/>
    <choice value="ouster_os2"/>
    <choice value="pandar_40p"/>
    <choice value="pandar_64"/>
    <choice value="pandar_qt"/>
    <choice value="pandar_xt32"/>
    <choice value="robosense_m1"/>
    <choice value="seyond_falcon"/>
    <choice value="seyond_robin_w"/>
    <choice value="velodyne_vlp16"/>
    <choice value="velodyne_vls128"/>
  </arg>

  <arg name="calibration_service_name" default="extrinsic_calibration"/>

  <!-- lidartag detector -->
  <include file="$(find-pkg-share lidartag)/launch/lidartag.launch.xml">
    <arg name="lidar_model" value="$(var lidar_model)"/>
    <arg name="pointcloud_topic" value="$(var pointcloud_topic)"/>
  </include>

  <!-- apriltag detector -->
  <include file="$(find-pkg-share tag_based_pnp_calibrator)/launch/apriltag_16h5.launch.py">
    <arg name="image_topic" value="$(var image_topic)"/>
    <arg name="camera_info_topic" value="$(var camera_info_topic)"/>
  </include>

  <!-- calibrator -->
  <node pkg="tag_based_pnp_calibrator" exec="tag_based_pnp_calibrator" name="tag_based_pnp_calibrator_node" output="screen">
    <param name="calib_rate" value="10.0"/>
    <param name="base_frame" value="$(var base_frame)"/>

    <!-- filtering parameters -->
    <param name="min_tag_size" value="0.6"/>
    <!--0.6 = 0.8 * 0.75 the apriltag size, not the lidartag-->
    <param name="max_tag_distance" value="20.0"/>
    <param name="max_allowed_homography_error" value="0.5"/>
    <param name="use_receive_time" value="$(var use_receive_time)"/>
    <param name="use_rectified_image" value="$(var use_rectified_image)"/>

    <param name="calibration_crossvalidation_training_ratio" value="0.7"/>
    <param name="calibration_convergence_min_pairs" value="$(var calibration_pairs)"/>
    <!-- the calibration finishes only after enough calibration pairs have been collected -->
    <param name="calibration_convergence_min_area_percentage" value="0.005"/>
    <param name="min_pnp_points" value="8"/>
    <param name="min_convergence_time" value="6.0"/>
    <param name="max_no_observation_time" value="3.0"/>
    <param name="new_hypothesis_distance" value="$(var calibration_pairs_min_distance)"/>
    <param name="tag_ids" value="[0, 1, 2, 3, 4, 5]"/>
    <param name="tag_sizes" value="[0.6, 0.6, 0.6, 0.6, 0.6, 0.6]"/>
    <!-- the sizes of the corresponding tag_ids - size of the black frame only  -->

    <param name="lidartag_max_convergence_translation" value="0.05"/>
    <param name="lidartag_max_convergence_translation_dot" value="0.03"/>
    <param name="lidartag_max_convergence_rotation" value="3.0"/>
    <param name="lidartag_max_convergence_rotation_dot" value="2.5"/>
    <param name="lidartag_new_hypothesis_translation" value="0.1"/>
    <param name="lidartag_new_hypothesis_rotation" value="15.0"/>
    <param name="lidartag_measurement_noise_translation" value="0.05"/>
    <param name="lidartag_measurement_noise_rotation" value="5.0"/>
    <param name="lidartag_process_noise_translation" value="0.01"/>
    <param name="lidartag_process_noise_translation_dot" value="0.001"/>
    <param name="lidartag_process_noise_rotation" value="1.0"/>
    <param name="lidartag_process_noise_rotation_dot" value="0.1"/>

    <param name="apriltag_max_convergence_translation" value="2.0"/>
    <param name="apriltag_new_hypothesis_translation" value="20.0"/>

    <param name="apriltag_measurement_noise_translation" value="0.2"/>
    <param name="apriltag_process_noise_translation" value="0.02"/>

    <remap from="image" to="$(var image_topic)"/>
    <remap from="camera_info" to="$(var camera_info_topic)"/>
    <remap from="extrinsic_calibration" to="$(var calibration_service_name)"/>
  </node>
</launch>
